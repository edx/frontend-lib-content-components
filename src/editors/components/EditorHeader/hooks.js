import React from 'react';
import * as module from './hooks';

export const hooks = {
  isEditing: () => {
    const [isEditing, setIsEditing] = React.useState(false);
    return {
      isEditing,
      startEditing: () => setIsEditing(true),
      stopEditing: () => setIsEditing(false),
    };
  },
  localTitle: ({
    setBlockTitle,
    stopEditing,
    typeHeader,
  }) => {
    const [localTitle, setLocalTitle] = React.useState(typeHeader);
    return {
      updateTitle: () => {
        setBlockTitle(localTitle);
        stopEditing();
      },
      handleChange: (e) => setLocalTitle(e.target.value),
      localTitle,
    };
  },
  handleKeyDown: ({ stopEditing, editorRef }) => (e) => {
    if (e.key === 'Enter') {
      stopEditing();
    }
    if (e.key === 'Tab' && editorRef) {
      e.preventDefault();
      editorRef.current.focus();
    }
  },
};

/* eslint-disable import/prefer-default-export */
export const localTitleHooks = ({
  editorRef,
  setBlockTitle,
  typeHeader,
}) => {
  const { isEditing, startEditing, stopEditing } = module.hooks.isEditing();
  const { localTitle, handleChange, updateTitle } = module.hooks.localTitle({
    setBlockTitle,
    stopEditing,
    typeHeader,
  });
  return {
    isEditing,
    startEditing,
    stopEditing,

    localTitle,
    updateTitle,
    handleChange,

    inputRef: React.createRef(),
    handleKeyDown: module.hooks.handleKeyDown({ stopEditing, editorRef }),
  };
};
