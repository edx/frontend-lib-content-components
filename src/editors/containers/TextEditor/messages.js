export const messages = {
  couldNotLoadTextContext: {
    id: 'authoring.texteditor.load.error',
    defaultMessage: 'Error: Could Not Load Text Content',
    description: 'Error Message Dispayed When HTML content fails to Load',
  },
};

export default messages;
