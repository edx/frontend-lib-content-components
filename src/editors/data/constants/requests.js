import { StrictDict } from '../../utils';

export const RequestStates = StrictDict({
  inactive: 'inactive',
  pending: 'pending',
  completed: 'completed',
  failed: 'failed',
});

export const RequestKeys = StrictDict({
  fetchBlock: 'fetchBlock',
  fetchUnit: 'fetchUnit',
  fetchStudioView: 'fetchStudioView',
  saveBlock: 'saveBlock',
  fetchImages: 'fetchImages',
  uploadImage: 'uploadImage',
});
