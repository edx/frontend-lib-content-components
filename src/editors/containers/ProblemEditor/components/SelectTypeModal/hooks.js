import {
  useState,
} from 'react';
import { StrictDict } from '../../utils';

import * as module from './hooks';

export const state = StrictDict({
  selected: (val) => useState(val),
});

export default { state };
