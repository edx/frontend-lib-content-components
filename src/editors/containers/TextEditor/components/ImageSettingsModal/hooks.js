import React from 'react';

import { StrictDict } from '../../../../utils';
import * as module from './hooks';

// Simple wrappers for useState to allow easy mocking for tests.
export const state = {
  dimensions: (val) => React.useState(val),
  locked: (val) => React.useState(val),
  local: (val) => React.useState(val),
  lockInitialized: (val) => React.useState(val),
  altText: (val) => React.useState(val),
  isDecorative: (val) => React.useState(val),
};

export const dimKeys = StrictDict({
  height: 'height',
  width: 'width',
});

/**
 * findGcd(numerator, denominator)
 * Find the greatest common denominator of a ratio or fraction.
 * @param {number} numerator - ratio numerator
 * @param {number} denominator - ratio denominator
 * @return {number} - ratio greatest common denominator
 */
export const findGcd = (a, b) => (b ? findGcd(b, a % b) : a);
const checkEqual = (d1, d2) => (d1.height === d2.height && d1.width === d2.width);

/**
 * getValidDimensions({ dimensions, local, locked })
 * Find valid ending dimensions based on start state, request, and lock state
 * @param {obj} dimensions - current stored dimensions
 * @param {obj} local - local (active) dimensions in the inputs
 * @param {obj} locked - locked dimensions
 * @return {obj} - output dimensions after move ({ height, width })
 */
export const getValidDimensions = ({
  dimensions,
  local,
  locked,
}) => {
  const out = {};
  let iter;
  const { minInc } = locked;
  const isMin = dimensions.height === minInc.height;

  const keys = (local.height !== dimensions.height)
    ? { changed: dimKeys.height, other: dimKeys.width }
    : { changed: dimKeys.width, other: dimKeys.height };

  const direction = local[keys.changed] > dimensions[keys.changed] ? 1 : -1;

  // don't move down if already at minimum size
  if (direction < 0 && isMin) { return dimensions; }
  // find closest valid iteration of the changed field
  iter = Math.max(Math.round(local[keys.changed] / minInc[keys.changed]), 1);
  // if closest valid iteration is current iteration, move one iteration in the change direction
  if (iter === (dimensions[keys.changed] / minInc[keys.changed])) { iter += direction; }

  out[keys.changed] = iter * minInc[keys.changed];
  out[keys.other] = out[keys.changed] * (locked[keys.other] / locked[keys.changed]);

  return out;
};

/**
 * newDimensions({ dimensions, local, locked })
 * Returns the local dimensions if unlocked or unchanged, and otherwise returns new valid
 * dimensions.
 * @param {obj} dimensions - current stored dimensions
 * @param {obj} local - local (active) dimensions in the inputs
 * @param {obj} locked - locked dimensions
 * @return {obj} - output dimensions after attempted move ({ height, width })
 */
export const newDimensions = ({ dimensions, local, locked }) => (
  (!locked || checkEqual(local, dimensions))
    ? local
    : module.getValidDimensions({ dimensions, local, locked })
);

/**
 * lockDimensions({ dimensions, lockInitialized, setLocked })
 * Lock dimensions if lock initialized.  Store minimum valid increment on lock so
 * that we don't have re-compute.
 * @param {obj} dimensions - current stored dimensions
 * @param {bool} lockInitialized - has the lock state initialized?
 * @param {func} setLocked - set lock state
 */
export const lockDimensions = ({ dimensions, lockInitialized, setLocked }) => {
  if (!lockInitialized) { return; }

  // find minimum viable increment
  let gcd = findGcd(dimensions.width, dimensions.height);
  if ([dimensions.width, dimensions.height].some(v => !Number.isInteger(v / gcd))) {
    gcd = 1;
  }
  const minInc = { width: dimensions.width / gcd, height: dimensions.height / gcd, gcd };
  setLocked({ ...dimensions, minInc });
};

/**
 * dimensionLockHooks({ dimensions })
 * Returns a set of hooks pertaining to the dimension locks.
 * Locks the dimensions initially, on lock initialization.
 * @param {obj} dimensions - current stored dimensions
 * @return {obj} - dimension lock hooks
 *   {func} initializeLock - enable the lock mechanism
 *   {obj} locked - current locked state
 *   {func} lock - lock the current dimensions
 *   {func} unlock - unlock the dimensions
 */
export const dimensionLockHooks = ({ dimensions }) => {
  const [locked, setLocked] = module.state.locked(null);
  const [lockInitialized, setLockInitialized] = module.state.lockInitialized(null);
  const lock = () => module.lockDimensions({ lockInitialized, dimensions, setLocked });

  React.useEffect(lock, [lockInitialized]);

  return {
    initializeLock: () => setLockInitialized(true),
    locked,
    lock,
    unlock: () => setLocked(null),
  };
};

/**
 * dimensionHooks()
 * Returns an object of dimension-focused react hooks.
 * @return {obj} - dimension hooks
 *   {func} onImgLoad - initializes image dimension fields
 *     @param {object} selection - selected image object with possible override dimensions.
 *     @return {callback} - image load event callback that loads dimensions.
 *   {object} locked - current locked state
 *   {func} lock - lock current dimensions
 *   {func} unlock - unlock dimensions
 *   {object} value - current dimension values
 *   {func} setHeight - set height
 *     @param {string} - new height string
 *   {func} setWidth - set width
 *     @param {string} - new width string
 *   {func} updateDimensions - set dimensions based on state
 */
export const dimensionHooks = () => {
  const [dimensions, setDimensions] = module.state.dimensions(null);
  const [local, setLocal] = module.state.local(null);
  const setAll = (value) => {
    setDimensions(value);
    setLocal(value);
  };
  const {
    initializeLock,
    lock,
    locked,
    unlock,
  } = module.dimensionLockHooks({ dimensions });
  return {
    onImgLoad: (selection) => ({ target: img }) => {
      setAll({
        height: selection.height || img.naturalHeight,
        width: selection.width || img.naturalWidth,
      });
      initializeLock();
    },
    locked,
    lock,
    unlock,
    value: local,
    setHeight: (height) => setLocal({ ...local, height: parseInt(height, 10) }),
    setWidth: (width) => setLocal({ ...local, width: parseInt(width, 10) }),
    updateDimensions: () => setAll(module.newDimensions({ dimensions, local, locked })),
  };
};

/**
 * altTextHooks(savedText)
 * Returns a set of react hooks focused around alt text
 * @return {obj} - alt text hooks
 *   {string} value - alt text value
 *   {func} setValue - set alt test value
 *     @param {string} - new alt text
 *   {bool} isDecorative - is the image decorative?
 *   {func} setIsDecorative - set isDecorative field
 *     @param {bool} isDecorative
 */
export const altTextHooks = (savedText) => {
  const [value, setValue] = module.state.altText(savedText || '');
  const [isDecorative, setIsDecorative] = module.state.isDecorative(false);
  return {
    value,
    setValue,
    isDecorative,
    setIsDecorative,
  };
};

/**
 * onInputChange(handleValue)
 * Simple event handler forwarding the event target value to a given callback
 * @param {func} handleValue - event value handler
 * @return {func} - evt callback that will call handleValue with the event target value.
 */
export const onInputChange = (handleValue) => (e) => handleValue(e.target.value);

/**
 * onCheckboxChange(handleValue)
 * Simple event handler forwarding the event target checked prop to a given callback
 * @param {func} handleValue - event value handler
 * @return {func} - evt callback that will call handleValue with the event target checked prop.
 */
export const onCheckboxChange = (handleValue) => (e) => handleValue(e.target.checked);

/**
 * onSave({ altText, dimensions, isDecorative, saveToEditor })
 * Handle saving the image context to the text editor
 * @param {string} altText - image alt text
 * @param {object} dimension - image dimensions ({ width, height })
 * @param {bool} isDecorative - is the image decorative?
 * @param {func} saveToEditor - save method for submitting image settings.
 */
export const onSaveClick = ({
  altText,
  dimensions,
  isDecorative,
  saveToEditor,
}) => () => saveToEditor({
  altText,
  dimensions,
  isDecorative,
});

/**
 * isSaveDisabled(altText)
 * Returns true the save button should be disabled (altText is missing and not decorative)
 * @param {object} altText - altText hook object
 *   {bool} isDecorative - is the image decorative?
 *   {string} value - alt text value
 * @return {bool} - should the save button be disabled?
 */
export const isSaveDisabled = (altText) => !altText.isDecorative && (altText.value === '');

export default {
  altText: altTextHooks,
  dimensions: dimensionHooks,
  isSaveDisabled,
  onCheckboxChange,
  onInputChange,
  onSaveClick,
};
