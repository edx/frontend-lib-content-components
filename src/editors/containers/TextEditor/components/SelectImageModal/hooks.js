import React from 'react';
import * as module from './hooks';
import { sortFunctions, sortKeys } from './utils';

export const state = {
  images: (val) => React.useState(val),
  highlighted: (val) => React.useState(val),
  searchString: (val) => React.useState(val),
  sortBy: (val) => React.useState(val),
};

export const searchAndSortHooks = () => {
  const [searchString, setSearchString] = module.state.searchString('');
  const [sortBy, setSortBy] = module.state.sortBy(sortKeys.dateNewest);
  return {
    searchString,
    onSearchChange: e => setSearchString(e.target.value),
    clearSearchString: () => setSearchString(''),
    sortBy,
    onSortClick: key => () => setSortBy(key),
  };
};

export const filteredList = ({ searchString, imageList }) => imageList.filter(
  ({ displayName }) => displayName.toLowerCase().includes(searchString.toLowerCase()),
);

export const displayList = ({ sortBy, searchString, images }) => module.filteredList({
  searchString,
  imageList: Object.values(images),
}).sort(sortFunctions[sortBy in sortKeys ? sortKeys[sortBy] : sortKeys.dateNewest]);

export const imgListHooks = ({
  fetchImages,
  setSelection,
  searchSortProps,
}) => {
  const [images, setImages] = module.state.images({});
  const [highlighted, setHighlighted] = module.state.highlighted(null);

  React.useEffect(() => {
    fetchImages({ onSuccess: setImages });
  }, []);

  return {
    images,
    // highlight by id
    selectBtnProps: {
      disabled: !highlighted,
      onClick: () => setSelection(images[highlighted]),
    },
    galleryProps: {
      highlighted,
      onHighlightChange: e => setHighlighted(e.target.value),
      displayList: module.displayList({ ...searchSortProps, images }),
    },
  };
};

export const fileInputHooks = ({ uploadImage }) => {
  const ref = React.useRef();
  const click = () => ref.current.click();
  const resetFile = () => { ref.current.value = ''; };
  const addFile = (e) => uploadImage({
    file: e.target.files[0],
    resetFile,
  });
  return {
    click,
    addFile,
    ref,
  };
};

export const imgHooks = ({ fetchImages, uploadImage, setSelection }) => {
  const searchSortProps = module.searchAndSortHooks();
  const imgList = module.imgListHooks({ fetchImages, setSelection, searchSortProps });
  const fileInput = module.fileInputHooks({ uploadImage });
  const { selectBtnProps, galleryProps } = imgList;

  return {
    fileInput,
    galleryProps,
    selectBtnProps,
    searchSortProps,
  };
};

export default {
  imgHooks,
};
