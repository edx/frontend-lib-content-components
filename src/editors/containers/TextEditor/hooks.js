import { useState } from 'react';

import { StrictDict } from '../../utils/index';
import * as module from './hooks';

export const state = {
  isModalOpen: (val) => useState(val),
  imageSelection: (val) => useState(val),
};

export const openModalWithSelectedImage = (editor, setImage, openModal) => () => {
  const imgHTML = editor.selection.getNode();
  setImage({
    externalUrl: imgHTML.src,
    altText: imgHTML.alt,
    width: imgHTML.width,
    height: imgHTML.height,
  });
  openModal();
};

export const addImageUploadBehavior = ({ openModal, setImage }) => (editor) => {
  editor.ui.registry.addButton('imageuploadbutton', {
    icon: 'image',
    onAction: openModal,
  });
  editor.ui.registry.addButton('editimagesettings', {
    icon: 'image',
    onAction: module.openModalWithSelectedImage(editor, setImage, openModal),
  });
};

export const initializeEditorRef = (setRef, initializeEditor) => (editor) => {
  setRef(editor);
  initializeEditor();
};

// for toast onClose to avoid console warnings
export const nullMethod = () => {};

export const pluginConfig = {
  plugins: StrictDict({
    link: 'link',
    codesample: 'codesample',
    emoticons: 'emoticons',
    table: 'table',
    charmap: 'charmap',
    code: 'code',
    autoresize: 'autoresize',
    image: 'image',
    imagetools: 'imagetools',
  }),
  menubar: false,
  toolbar: StrictDict({
    do: 'undo redo',
    formatselect: 'formatselect',
    wieght: 'bold italic backcolor',
    align: 'alignleft aligncenter alignright alignjustify',
    indents: 'bullist numlist outdent indent ',
    imageupload: 'imageuploadbutton',
    link: 'link',
    emoticons: 'emoticons',
    table: 'table',
    codesample: 'codesample',
    charmap: 'charmap',
    removeformat: 'removeformat',
    hr: 'hr',
    code: 'code',
  }),
  imageToolbar: StrictDict({
    rotate: 'rotateleft rotateright',
    flip: 'flipv fliph',
    editImageSettings: 'editimagesettings',
  }),
};
export const getConfig = (key) => {
  if (key === 'imageToolbar' || key === 'toolbar') {
    return Object.values(module.pluginConfig[key]).join(' | ');
  }
  return Object.values(module.pluginConfig[key]).join(' ');
};

export const editorConfig = ({
  setEditorRef,
  blockValue,
  openModal,
  initializeEditor,
  setSelection,
}) => ({
  onInit: (evt, editor) => module.initializeEditorRef(setEditorRef, initializeEditor)(editor),
  initialValue: blockValue ? blockValue.data.data : '',
  init: {
    setup: module.addImageUploadBehavior({ openModal, setImage: setSelection }),
    plugins: module.getConfig('plugins'),
    menubar: false,
    toolbar: module.getConfig('toolbar'),
    imagetools_toolbar: module.getConfig('imageToolbar'),
    imagetools_cors_hosts: ['courses.edx.org'],
    height: '100%',
    min_height: 1000,
    branding: false,
  },
});

export const selectedImage = (val) => {
  const [selection, setSelection] = module.state.imageSelection(val);
  return {
    clearSelection: () => setSelection(null),
    selection,
    setSelection,
  };
};

export const modalToggle = () => {
  const [isOpen, setIsOpen] = module.state.isModalOpen(false);
  return {
    isOpen,
    openModal: () => setIsOpen(true),
    closeModal: () => setIsOpen(false),
  };
};
