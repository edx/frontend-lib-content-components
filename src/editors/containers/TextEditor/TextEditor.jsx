import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Editor } from '@tinymce/tinymce-react';

import 'tinymce';
import 'tinymce/themes/silver';
import 'tinymce/skins/ui/oxide/skin.css';
import 'tinymce/icons/default';
import 'tinymce/plugins/link';
import 'tinymce/plugins/table';
import 'tinymce/plugins/codesample';
import 'tinymce/plugins/emoticons';
import 'tinymce/plugins/emoticons/js/emojis';
import 'tinymce/plugins/charmap';
import 'tinymce/plugins/code';
import 'tinymce/plugins/autoresize';
import 'tinymce/plugins/image';
import 'tinymce/plugins/imagetools';

import { FormattedMessage } from '@edx/frontend-platform/i18n';
import {
  Spinner,
  Toast,
} from '@edx/paragon';

import { actions, selectors } from '../../data/redux';
import { RequestKeys } from '../../data/constants/requests';
import {
  editorConfig,
  modalToggle,
  nullMethod,
  selectedImage,
} from './hooks';
import ImageUploadModal from './components/ImageUploadModal';
import messages from './messages';

export const TextEditor = ({
  setEditorRef,
  editorRef,
  // redux
  blockValue,
  blockFailed,
  blockFinished,
  initializeEditor,
}) => {
  const { isOpen, openModal, closeModal } = modalToggle();

  // selected image file reference data object.
  // this field determines the step of the ImageUploadModal
  const imageSelection = selectedImage(null);

  return (
    <div className="editor-body h-75">
      <ImageUploadModal
        isOpen={isOpen}
        close={closeModal}
        editorRef={editorRef}
        {...imageSelection}
      />

      <Toast show={blockFailed} onClose={nullMethod}>
        <FormattedMessage {...messages.couldNotLoadTextContext} />
      </Toast>

      {(!blockFinished)
        ? (
          <div className="text-center p-6">
            <Spinner animation="border" className="m-3" screenreadertext="loading" />
          </div>
        )
        : (
          <Editor
            {...editorConfig({
              setEditorRef,
              blockValue,
              openModal,
              initializeEditor,
              setSelection: imageSelection.setSelection,
              clearSelection: imageSelection.clearSelection,
            })}
          />
        )}
    </div>
  );
};
TextEditor.defaultProps = {
  blockValue: null,
  editorRef: null,
};
TextEditor.propTypes = {
  editorRef: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ current: PropTypes.any }),
  ]),
  setEditorRef: PropTypes.func.isRequired,
  // redux
  blockValue: PropTypes.shape({
    data: PropTypes.shape({ data: PropTypes.string }),
  }),
  blockFailed: PropTypes.bool.isRequired,
  blockFinished: PropTypes.bool.isRequired,
  initializeEditor: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => ({
  blockValue: selectors.app.blockValue(state),
  blockFailed: selectors.requests.isFailed(state, { requestKey: RequestKeys.fetchBlock }),
  blockFinished: selectors.requests.isFinished(state, { requestKey: RequestKeys.fetchBlock }),
});

export const mapDispatchToProps = {
  initializeEditor: actions.app.initializeEditor,
};

export default connect(mapStateToProps, mapDispatchToProps)(TextEditor);
