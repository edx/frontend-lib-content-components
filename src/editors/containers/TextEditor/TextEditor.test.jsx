import React from 'react';
import { shallow } from 'enzyme';
import { TextEditor, mapStateToProps, mapDispatchToProps } from './TextEditor';
import { actions, selectors } from '../../data/redux';
import { RequestKeys } from '../../data/constants/requests';
import { modalToggle } from './hooks';

// Per https://github.com/tinymce/tinymce-react/issues/91 React unit testing in JSDOM is not supported by tinymce.
// Consequently, mock the Editor out.
jest.mock('@tinymce/tinymce-react', () => {
  const originalModule = jest.requireActual('@tinymce/tinymce-react');
  return {
    __esModule: true,
    ...originalModule,
    Editor: () => 'TiNYmCE EDitOR',
  };
});

jest.mock('./components/ImageUploadModal', () => 'ImageUploadModal');

jest.mock('./hooks', () => {
  const updateState = jest.fn();
  return ({
    editorConfig: jest.fn(args => ({ editorConfig: args })),
    modalToggle: jest.fn(() => ({
      isOpen: true,
      openModal: jest.fn().mockName('openModal'),
      closeModal: jest.fn().mockName('closeModal'),
    })),
    selectedImage: jest.fn(val => ({
      selection: 'hooks.selectedImage.selection',
      setSelection: jest.fn().mockName('hooks.selectedImage.setSelection'),
      clearSelection: jest.fn().mockName('hooks.selectedImage.clearSelection'),
    })),
    nullMethod: jest.fn().mockName('hooks.nullMethod'),
  });
});

jest.mock('react', () => {
  const updateState = jest.fn();
  return {
    ...jest.requireActual('react'),
    updateState,
    useState: jest.fn(val => ([{ state: val }, jest.fn().mockName('setState')])),
  };
});

jest.mock('../../data/redux', () => ({
  actions: {
    app: {
      initializeEditor: jest.fn().mockName('actions.app.initializeEditor'),
    },
  },
  selectors: {
    app: {
      blockValue: jest.fn(state => ({ blockValue: state })),
    },
    requests: {
      isFailed: jest.fn((state, params) => ({ isFailed: { state, params } })),
      isFinished: jest.fn((state, params) => ({ isFailed: { state, params } })),
    },
  },
}));

describe('TextEditor', () => {
  const props = {
    setEditorRef: jest.fn().mockName('args.setEditorRef'),
    editorRef: { current: { value: 'something' } },
    // redux
    blockValue: { data: { some: 'eDiTablE Text' } },
    blockFailed: false,
    blockFinished: true,
    initializeEditor: jest.fn().mockName('args.intializeEditor'),
  };
  describe('snapshots', () => {
    modalToggle.mockReturnValue({
      isOpen: false,
      openModal: jest.fn().mockName('modal.openModal'),
      closeModal: jest.fn().mockName('modal.closeModal'),
    });
    test('renders as expected with default behavior', () => {
      expect(shallow(<TextEditor {...props} />)).toMatchSnapshot();
    });
    test('not yet loaded, Spinner appears', () => {
      expect(shallow(<TextEditor {...props} blockFinished={false} />)).toMatchSnapshot();
    });
    test('block failed to load, Toast is shown', () => {
      expect(shallow(<TextEditor {...props} blockFailed />)).toMatchSnapshot();
    });
  });
  describe('mapStateToProps', () => {
    const testState = { A: 'pple', B: 'anana', C: 'ucumber' };
    test('blockValue from app.blockValue', () => {
      expect(
        mapStateToProps(testState).blockValue,
      ).toEqual(selectors.app.blockValue(testState));
    });
    test('blockFailed from requests.isFailed', () => {
      expect(
        mapStateToProps(testState).blockFailed,
      ).toEqual(selectors.requests.isFailed(testState, { requestKey: RequestKeys.fetchBlock }));
    });
    test('blockFinished from requests.isFinished', () => {
      expect(
        mapStateToProps(testState).blockFinished,
      ).toEqual(selectors.requests.isFinished(testState, { requestKey: RequestKeys.fetchBlock }));
    });
  });
  describe('mapDispatchToProps', () => {
    test('initializeEditor from actions.app.initializeEditor', () => {
      expect(mapDispatchToProps.initializeEditor).toEqual(actions.app.initializeEditor);
    });
  });
});
