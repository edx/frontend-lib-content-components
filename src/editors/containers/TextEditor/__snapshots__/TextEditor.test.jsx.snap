// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextEditor snapshots block failed to load, Toast is shown 1`] = `
<div
  className="editor-body h-75"
>
  <ImageUploadModal
    clearSelection={[MockFunction hooks.selectedImage.clearSelection]}
    close={[MockFunction modal.closeModal]}
    editorRef={
      Object {
        "current": Object {
          "value": "something",
        },
      }
    }
    isOpen={false}
    selection="hooks.selectedImage.selection"
    setSelection={[MockFunction hooks.selectedImage.setSelection]}
  />
  <Toast
    onClose={[MockFunction hooks.nullMethod]}
    show={true}
  >
    <FormattedMessage
      defaultMessage="Error: Could Not Load Text Content"
      description="Error Message Dispayed When HTML content fails to Load"
      id="authoring.texteditor.load.error"
    />
  </Toast>
  <Editor
    editorConfig={
      Object {
        "blockValue": Object {
          "data": Object {
            "some": "eDiTablE Text",
          },
        },
        "clearSelection": [MockFunction hooks.selectedImage.clearSelection],
        "initializeEditor": [MockFunction args.intializeEditor],
        "openModal": [MockFunction modal.openModal],
        "setEditorRef": [MockFunction args.setEditorRef],
        "setSelection": [MockFunction hooks.selectedImage.setSelection],
      }
    }
  />
</div>
`;

exports[`TextEditor snapshots not yet loaded, Spinner appears 1`] = `
<div
  className="editor-body h-75"
>
  <ImageUploadModal
    clearSelection={[MockFunction hooks.selectedImage.clearSelection]}
    close={[MockFunction modal.closeModal]}
    editorRef={
      Object {
        "current": Object {
          "value": "something",
        },
      }
    }
    isOpen={false}
    selection="hooks.selectedImage.selection"
    setSelection={[MockFunction hooks.selectedImage.setSelection]}
  />
  <Toast
    onClose={[MockFunction hooks.nullMethod]}
    show={false}
  >
    <FormattedMessage
      defaultMessage="Error: Could Not Load Text Content"
      description="Error Message Dispayed When HTML content fails to Load"
      id="authoring.texteditor.load.error"
    />
  </Toast>
  <div
    className="text-center p-6"
  >
    <Spinner
      animation="border"
      className="m-3"
      screenreadertext="loading"
    />
  </div>
</div>
`;

exports[`TextEditor snapshots renders as expected with default behavior 1`] = `
<div
  className="editor-body h-75"
>
  <ImageUploadModal
    clearSelection={[MockFunction hooks.selectedImage.clearSelection]}
    close={[MockFunction modal.closeModal]}
    editorRef={
      Object {
        "current": Object {
          "value": "something",
        },
      }
    }
    isOpen={false}
    selection="hooks.selectedImage.selection"
    setSelection={[MockFunction hooks.selectedImage.setSelection]}
  />
  <Toast
    onClose={[MockFunction hooks.nullMethod]}
    show={false}
  >
    <FormattedMessage
      defaultMessage="Error: Could Not Load Text Content"
      description="Error Message Dispayed When HTML content fails to Load"
      id="authoring.texteditor.load.error"
    />
  </Toast>
  <Editor
    editorConfig={
      Object {
        "blockValue": Object {
          "data": Object {
            "some": "eDiTablE Text",
          },
        },
        "clearSelection": [MockFunction hooks.selectedImage.clearSelection],
        "initializeEditor": [MockFunction args.intializeEditor],
        "openModal": [MockFunction modal.openModal],
        "setEditorRef": [MockFunction args.setEditorRef],
        "setSelection": [MockFunction hooks.selectedImage.setSelection],
      }
    }
  />
</div>
`;
